homeassistant/util/logging.py

```diff
@@ -164,3 +164,46 @@ def wrapper(*args: Any) -> None:
                 log_exception(*args)
         wrapper_func = wrapper
     return wrapper_func
+
+
+def catch_log_coro_exception(
+        func: Coroutine[Any, Any, Any],
+        format_err: Callable[..., Any],
+        *args: Any) -> Coroutine[Any, Any, Any]:
+    """Decorate a coroutine to catch and log exceptions."""
+    def log_exception(*args: Any) -> None:
+        module_name = inspect.getmodule(inspect.trace()[1][0]).__name__
+        # Do not print the wrapper in the traceback
+        frames = len(inspect.trace()) - 1
+        exc_msg = traceback.format_exc(-frames)
+        friendly_msg = format_err(*args)
+        logging.getLogger(module_name).error('%s\n%s', friendly_msg, exc_msg)
+
+    @wraps(func)  # type: ignore
+    async def coro_wrapper(*args: Any) -> Any:
+        """Catch and log exception."""
+        try:
+            return await func
+        except Exception:  # pylint: disable=broad-except
+            log_exception(*args)
+            return None
+    return coro_wrapper()  # type: ignore
```

Comment:
MyPy doesn't think we're returning a coroutine, possible because of @wrap:
`warning: Returning Any from function declared to return "Coroutine[Any, Any, Any]"`

Reply:
You can use `cast`

Reply:
The problem was the incorrect use of `@wrap`